scalableminds/webknossos
# Annotation Locking Mechanism

new route `POST /api/annotations/<annotationId>/acquireMutex` (no body) will, for othersMayEdit annotations, return json body with two fields: canEdit: bool, blockedByUser: user (optional, only set if blocked. Uses compact version of user json, just like in user/owner in annotation json)

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- create an additional user account / promote sample user2 to also be an admin.
- create an annotation with sample user 1, make it editable by others keep it open.
- Open the annotation in a private tab with sample user2. There it should switch to read only and show a toast that sample user is currently editing the annotation
- When you close the annotation with sample user and wait for max 3 minutes, the annotation should be available to sample user2 and automatically reload the annotation once the mutex is acquired.
- Now reopen the annotation with sample user and try to turn of `othersMayEdit`. This should not be possible, as sample user2 currently has the mutex.
- close the annotation with sample user2
- wait 3 mins and sample user should now be able to edit the annotation and the `othersMayEdit` setting again.#
- and of cause: Play around with it the way you want to discover uncovered edge cases :D

### Issues:
- fixes #6800

------
